v2/pkg/genruntime/extensions/postreconciliation_checker.go

```diff
@@ -0,0 +1,160 @@
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+)
+
+// PostReconciliationChecker is implemented by resources that want to do extra status checks after
+// a full ARM reconcile.
+type PostReconciliationChecker interface {
+	// PostReconcileCheck does a post-reconcile check to see if the resource is in a state to set 'Ready' condition.
+	// ARM resources should implement this to avoid reconciliation attempts that cannot possibly succeed.
+	// Returns PostReconcileCheckResultSuccess if the reconciliation is successful.
+	// Returns PostReconcileCheckResultFailure and a human-readable reason if the reconciliation should put a condition on resource.
+	// ctx is the current operation context.
+	// obj is the resource about to be reconciled. The resource's State will be freshly updated.
+	// kubeClient allows access to the cluster for any required queries.
+	// armClient allows access to ARM for any required queries.
+	// log is the logger for the current operation.
+	PostReconcileCheck(
+		ctx context.Context,
+		obj genruntime.MetaObject,
+		owner genruntime.MetaObject,
+		kubeClient kubeclient.Client,
+		armClient *genericarmclient.GenericClient,
+		log logr.Logger,
+		next PostReconcileCheckFunc,
+	) (PostReconcileCheckResult, error)
+}
+
+type PostReconcileCheckFunc func(
+	ctx context.Context,
+	obj genruntime.MetaObject,
+	owner genruntime.MetaObject,
+	kubeClient kubeclient.Client,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger,
+) (PostReconcileCheckResult, error)
+
+type PostReconcileCheckResult struct {
+	action   postReconcileCheckResultType
+	severity conditions.ConditionSeverity
+	reason   conditions.Reason
+	message  string
+}
+
+// PostReconcileCheckResultSuccess indicates that a resource is ready after reconciliation by returning a PostReconcileCheckResult
+// with action `Success`.
+func PostReconcileCheckResultSuccess() PostReconcileCheckResult {
+	return PostReconcileCheckResult{
+		action: postReconcileCheckResultTypeSuccess,
+	}
+}
+
+// PostReconcileCheckResultFailure indicates post reconciliation check of a resource is currently failed by returning a PostReconcileCheckResult
+// with action `Failure`.
+// reason is an explanatory reason to show to the user via a warning condition on the resource.
+func PostReconcileCheckResultFailure(reason string) PostReconcileCheckResult {
+	return PostReconcileCheckResult{
+		action:   postReconcileCheckResultTypeFailure,
+		severity: conditions.ConditionSeverityWarning,
+		reason:   conditions.ReasonPostReconcileFailure,
+		message:  reason,
+	}
+}
+
+func (r PostReconcileCheckResult) ReconciliationFailure() bool {
```

Comment:
Suggest renaming these to use the past tense, since that matches the semantics.

```suggestion
func (r PostReconcileCheckResult) ReconciliationFailed() bool {
```